server/src/main/java/org/elasticsearch/index/mapper/KeyedJsonAtomicFieldData.java

```diff
@@ -0,0 +1,230 @@
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.index.mapper;
+
+import org.apache.lucene.index.SortedSetDocValues;
+import org.apache.lucene.util.Accountable;
+import org.apache.lucene.util.BytesRef;
+import org.elasticsearch.index.fielddata.AbstractSortedSetDocValues;
+import org.elasticsearch.index.fielddata.AtomicOrdinalsFieldData;
+import org.elasticsearch.index.fielddata.FieldData;
+import org.elasticsearch.index.fielddata.ScriptDocValues;
+import org.elasticsearch.index.fielddata.SortedBinaryDocValues;
+import org.elasticsearch.index.fielddata.plain.AbstractAtomicOrdinalsFieldData;
+
+import java.io.IOException;
+import java.util.Collection;
+
+/**
+ * The atomic field data implementation for {@link JsonFieldMapper.KeyedJsonFieldType}.
+ *
+ * This class wraps the field data that is built directly on the keyed JSON field,
+ * and filters out values whose prefix doesn't match the requested key.
+ */
+public class KeyedJsonAtomicFieldData implements AtomicOrdinalsFieldData {
+
+    private final String key;
+    private final AtomicOrdinalsFieldData delegate;
+
+    KeyedJsonAtomicFieldData(String key,
+                             AtomicOrdinalsFieldData delegate) {
+        this.key = key;
+        this.delegate = delegate;
+    }
+
+    @Override
+    public long ramBytesUsed() {
+        return delegate.ramBytesUsed();
+    }
+
+    @Override
+    public Collection<Accountable> getChildResources() {
+        return delegate.getChildResources();
+    }
+
+    @Override
+    public SortedSetDocValues getOrdinalsValues() {
+        SortedSetDocValues values = delegate.getOrdinalsValues();
+        return KeyedJsonDocValues.create(key, values);
+    }
+
+    @Override
+    public void close() {
+        delegate.close();
+    }
+
+    @Override
+    public ScriptDocValues<?> getScriptValues() {
+        return AbstractAtomicOrdinalsFieldData.DEFAULT_SCRIPT_FUNCTION
+            .apply(getOrdinalsValues());
+    }
+
+    @Override
+    public SortedBinaryDocValues getBytesValues() {
+        return FieldData.toString(getOrdinalsValues());
+    }
+
+    static class KeyedJsonDocValues extends AbstractSortedSetDocValues {
+
+        private final BytesRef key;
+        private final SortedSetDocValues delegate;
+
+        /**
+         * The first and last ordinals whose term has 'key' as a prefix.
+         */
+        private final long minOrd;
+        private final long maxOrd;
+
+        public static KeyedJsonDocValues create(String key, SortedSetDocValues delegate) {
+            BytesRef keyBytes = new BytesRef(key);
+
+            long minOrd, maxOrd;
+            try {
+                minOrd = findMinOrd(keyBytes, delegate);
+                maxOrd = minOrd != -1 ? findMaxOrd(keyBytes, delegate) : -1;
+            } catch (IOException e) {
+                throw new RuntimeException(e);
```

Comment:
UncheckedIOException would be a better fit.

Reply:
👍